File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.java
class Solution {

    long gcd(long a, long b) {
        while (b != 0) {
            long temp = a % b;
            a = b;
            b = temp;
        }
        return a;
    }

    long lcm(long a, long b) {
        return a / gcd(a, b) * b;
    }

    long count(long x, int[] coins) {
        long ans = 0;
        int n = coins.length;

        for (int mask = 1; mask < (1 << n); mask++) {
            long L = 1;
            int bits = 0;
            boolean valid = true;

            for (int i = 0; i < n; i++) {
                if ((mask & (1 << i)) != 0) {
                    bits++;

                    L = lcm(L, coins[i]);

                    if (L > x) {
                        valid = false;
                        break;
                    }
                }
            }

            if (!valid) continue;

            long ways = x / L;

            if (bits % 2 == 1)
                ans += ways;
            else
                ans -= ways;
        }

        return ans;
    }

    public long findKthSmallest(int[] coins, int k) {
        long minCoin = coins[0];

        for (int coin : coins) {
            minCoin = Math.min(minCoin, coin);
        }

        long low = 1;
        long high = minCoin * (long) k;

        while (low < high) {
            long mid = low + (high - low) / 2;

            if (count(mid, coins) >= k)
                high = mid;
            else
                low = mid + 1;
        }

        return low;
    }
}

File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.js
var findKthSmallest = function(coins, k) {

    function gcd(a, b) {
        while (b !== 0n) {
            let temp = a % b;
            a = b;
            b = temp;
        }

        return a;
    }

    function lcm(a, b) {
        return (a / gcd(a, b)) * b;
    }

    function count(x) {
        let ans = 0n;
        let n = coins.length;

        for (let mask = 1; mask < (1 << n); mask++) {

            let L = 1n;
            let bits = 0;
            let valid = true;

            for (let i = 0; i < n; i++) {

                if (mask & (1 << i)) {
                    bits++;

                    L = lcm(L, BigInt(coins[i]));

                    if (L > x) {
                        valid = false;
                        break;
                    }
                }
            }

            if (!valid)
                continue;

            let ways = x / L;

            if (bits % 2 === 1)
                ans += ways;
            else
                ans -= ways;
        }

        return ans;
    }

    let minCoin = Math.min(...coins);

    let low = 1n;
    let high = BigInt(minCoin) * BigInt(k);

    while (low < high) {

        let mid = low + (high - low) / 2n;

        if (count(mid) >= BigInt(k))
            high = mid;
        else
            low = mid + 1n;
    }

    return Number(low);
};